py/map.c

```diff
@@ -126,6 +126,26 @@ STATIC void mp_map_rehash(mp_map_t *map) {
     m_del(mp_map_elem_t, old_table, old_alloc);
 }
 
+
+#if MICROPY_OPT_MAP_LOOKUP_CACHE
+// This provides a cache of index to the last known position of that index in
+// any map. On a cache hit, this allows short-circuiting the full linear
+// search in the case of an ordered map (i.e. all builtin modules and
+// objects' locals dicts), and computation of the hash (and potentially some
+// linear probing) in the case of a regular map. Note the same cache is
+// shared across all maps.
+uint8_t map_lookup_cache[MICROPY_OPT_MAP_LOOKUP_CACHE_SIZE];
```

Comment:
IMO this should go in `py/mpstate.h`, in `mp_state_vm_t`, outside the root-pointer section of that struct.  Reasoning:
1. keeps all state localised to that file
2. if there were multiple VM instances in the one address space, then I'd argue that each VM should have its own table because they run totally independent code

It shouldn't impact code size or runtime (it should compile to the same machine instructions...).

BTW, would there be any issue with GIL-less multithreading (eg rp2) accessing this table?  My guess is no.

Reply:
Done -- branch updated.

Interestingly this did gain a small performance improvement. Possibly the new location makes it better for the data cache?

```
$ ./run-perfbench.py -s ~/lookup-global-pybv11 ~/lookup-state-pybv11
diff of scores (higher is better)
N=100 M=100              /home/jimmo/lookup-global-pybv11 -> /home/jimmo/lookup-state-pybv11         diff      diff% (error%)
bm_chaos.py                  361.52 ->     372.92 :     +11.40 =  +3.153% (+/-0.00%)
bm_fannkuch.py                78.19 ->      78.97 :      +0.78 =  +0.998% (+/-0.01%)
bm_fft.py                   2571.66 ->    2569.91 :      -1.75 =  -0.068% (+/-0.00%)
bm_float.py                 5991.20 ->    6168.25 :    +177.05 =  +2.955% (+/-0.01%)
bm_hexiom.py                  43.92 ->      44.22 :      +0.30 =  +0.683% (+/-0.00%)
bm_nqueens.py               4439.64 ->    4444.68 :      +5.04 =  +0.114% (+/-0.00%)
bm_pidigits.py               647.20 ->     651.21 :      +4.01 =  +0.620% (+/-0.38%)
misc_aes.py                  422.65 ->     425.99 :      +3.34 =  +0.790% (+/-0.01%)
misc_mandel.py              3559.81 ->    3598.03 :     +38.22 =  +1.074% (+/-0.01%)
misc_pystone.py             2400.13 ->    2429.40 :     +29.27 =  +1.220% (+/-0.01%)
misc_raytrace.py             381.46 ->     391.24 :      +9.78 =  +2.564% (+/-0.01%)
```

This is nice because the net effect of this PR is now +ve on all tests on PYBV11.

```
$ ./run-perfbench.py -s ~/3.cache-baseline-pybv11 ~/lookup-state-pybv11
diff of scores (higher is better)
N=100 M=100              /home/jimmo/3.cache-baseline-pybv11 -> /home/jimmo/lookup-state-pybv11         diff      diff% (error%)
bm_chaos.py                  320.05 ->     372.92 :     +52.87 = +16.519% (+/-0.02%)
bm_fannkuch.py                78.96 ->      78.97 :      +0.01 =  +0.013% (+/-0.01%)
bm_fft.py                   2545.01 ->    2569.91 :     +24.90 =  +0.978% (+/-0.00%)
bm_float.py                 5088.98 ->    6168.25 :   +1079.27 = +21.208% (+/-0.01%)
bm_hexiom.py                  35.64 ->      44.22 :      +8.58 = +24.074% (+/-0.00%)
bm_nqueens.py               4194.71 ->    4444.68 :    +249.97 =  +5.959% (+/-0.00%)
bm_pidigits.py               649.25 ->     651.21 :      +1.96 =  +0.302% (+/-0.31%)
misc_aes.py                  384.51 ->     425.99 :     +41.48 = +10.788% (+/-0.01%)
misc_mandel.py              3112.82 ->    3598.03 :    +485.21 = +15.587% (+/-0.01%)
misc_pystone.py             1968.33 ->    2429.40 :    +461.07 = +23.424% (+/-0.01%)
misc_raytrace.py             318.19 ->     391.24 :     +73.05 = +22.958% (+/-0.01%)
```

And overall a pretty good result compared to bytecode caching

```
$ ./run-perfbench.py -s ~/cache-bytecode-pybv11 ~/lookup-state-pybv11
diff of scores (higher is better)
N=100 M=100              /home/jimmo/cache-bytecode-pybv11 -> /home/jimmo/lookup-state-pybv11         diff      diff% (error%)
bm_chaos.py                  358.90 ->     372.92 :     +14.02 =  +3.906% (+/-0.01%)
bm_fannkuch.py                78.48 ->      78.97 :      +0.49 =  +0.624% (+/-0.01%)
bm_fft.py                   2576.19 ->    2569.91 :      -6.28 =  -0.244% (+/-0.00%)
bm_float.py                 6296.83 ->    6168.25 :    -128.58 =  -2.042% (+/-0.01%)
bm_hexiom.py                  34.45 ->      44.22 :      +9.77 = +28.360% (+/-0.00%)
bm_nqueens.py               4069.66 ->    4444.68 :    +375.02 =  +9.215% (+/-0.00%)
bm_pidigits.py               652.31 ->     651.21 :      -1.10 =  -0.169% (+/-0.36%)
misc_aes.py                  453.34 ->     425.99 :     -27.35 =  -6.033% (+/-0.01%)
misc_mandel.py              2960.68 ->    3598.03 :    +637.35 = +21.527% (+/-0.01%)
misc_pystone.py             2513.17 ->    2429.40 :     -83.77 =  -3.333% (+/-0.01%)
misc_raytrace.py             372.69 ->     391.24 :     +18.55 =  +4.977% (+/-0.01%)
```